iml-wasm-components/iml-action-dropdown/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+#[derive(serde::Deserialize, serde::Serialize, Debug, Eq, PartialEq, Hash, Clone)]
+pub struct LockChange {
+    job_id: u64,
+    content_type_id: u64,
+    item_id: u64,
+    description: String,
+    lock_type: LockType,
+    action: Action,
+}
+
+// Model
+
+#[derive(Clone)]
+struct Model {
+    records: Vec<Record>,
+    locks: Locks,
+    open: bool,
+}
+
+// Setup a default here, for initialization later.
+impl Default for Model {
+    fn default() -> Self {
+        Self {
+            records: vec![],
+            locks: HashMap::new(),
+            open: false,
+        }
+    }
+}
+
+// Update
+
+#[derive(Clone)]
+enum Msg {
+    Open(bool),
+    Locks(Locks)
+}
+
+/// The sole source of updating the model; returns a fresh one.
+fn update(msg: Msg, model: Model) -> Update<Model> {
+    match msg {
+        Msg::Open(open) => Render(Model { open, ..model }),
+        Msg::Locks(locks) => Render(Model { locks, ..model })
+    }
+}
+
+// View
+fn show_class(model: &Model) -> &str {
+    if model.open {
+        "show"
+    } else {
+        ""
+    }
+}
+
+/// The top-level component we pass to the virtual dom.
+fn view(state: seed::App<Msg, Model>, model: &Model) -> El<Msg> {
+    // Attrs, Style, Events, and children may be defined separately.
+    let outer_style = style!{
+            "display" => "flex";
+            "flex-direction" => "column";
+            "text-align" => "center"
+    };
+
+    let next_open = Msg::Open(!model.open);
+
+    let recordEls: Vec<El<Msg>> = model.records.clone().into_iter()
+        .map(|x| div![
+                    format!("{:?}", x)
+                ])
+                .collect();
+
+    div![
+        outer_style,
+            recordEls,
+        div![
+            class!["btn-group", show_class(&model)],
+            button![
+                class!["btn", "btn-primary", "dropdown-toggle"],
+                mouse_ev("click", move |ev| {
+                    ev.stop_propagation();
+                    ev.prevent_default();
+                    next_open.clone()
+                }),
+                "Actions"
+            ],
+            div![
+                class!["dropdown-menu", show_class(&model)],
+                a![attrs! {"href" => "#"}, class!["dropdown-item"], "foo"],
+                a![attrs! {"href" => "#"}, class!["dropdown-item"], "bar"],
+                div![attrs! {"href" => "#"}, class!["dropdown-divider"]],
+                a![attrs! {"href" => "#"}, class!["dropdown-item"], "baz"]
+            ]
+        ],
+    ]
+}
+
+fn window_events(model: Model) -> Vec<seed::Listener<Msg>> {
+    vec![mouse_ev("click", move |ev| Msg::Open(false))]
+}
+
+#[wasm_bindgen]
+pub fn render(x:&JsValue) {
+    let records:Vec<Record> = x.into_serde().unwrap();
+
+    let model = Model {
+        records,
+        ..Default::default()
+    };
+
+    let app = seed::App::build(model, update, view)
+        .window_events(window_events)
+        .mount("iml-add")
```

Comment:
I was just reading up on `mount` and it looks like it will render the app to the element holding an id of `iml-add`. By definition we can only have one element with that id. It may not affect our stuff but it's probably something to pay attention to. Do you think it's something to be concerned with?

Reply:
From https://seed-rs.org/guide/2:

> This will render your app to the element holding the id you passed; in the case of this example, “main”. The only part of the web page Seed will interact with is that element, so you can use other HTML not part of Seed, or other JS code/frameworks in the same document.

Reply:
This won't look like this in final code.

We will need to generate unique id's for each call and add them as the id.

Reply:
ah so it will be dynamic. That makes sense.